File: python/layer.py
# -*- coding: utf-8 -*-

from pyelemental import RESTObject
from pyelemental import validate_string_in_list, validate_float_in_list, validate_int_in_list, validate_required_int, validate_required_float, validate_required_string, validate_required_time, validate_maximum_float, validate_minimum_float, validate_maximum_int, validate_minimum_int, validate_maximum_length, validate_minimum_length, validate_pattern


class Layer(RESTObject):
    """ Represents a Layer in the 

        Notes:
            Layer of a docker image
    """

    def __init__(self, **kwargs):
        """ Initializes a Layer instance

          Notes:
              You can specify all parameters while calling this methods.
              A special argument named `data` will enable you to load the
              object from a Python dictionary

          Examples:
              >>> layer = Layer(id=u'xxxx-xxx-xxx-xxx', name=u'Layer')
              >>> layer = Layer(data=my_dict)
        """

        super(Layer, self).__init__()

        # Read/Write Attributes
        
        self._id = None
        self._name = None
        self._namespacename = None
        self._parentname = None
        self._severity = None
        self._vulnerabilities = None
        
        self.expose_attribute(local_name="id", remote_name="ID")
        self.expose_attribute(local_name="name", remote_name="name")
        self.expose_attribute(local_name="namespacename", remote_name="namespaceName")
        self.expose_attribute(local_name="parentname", remote_name="parentName")
        self.expose_attribute(local_name="severity", remote_name="severity")
        self.expose_attribute(local_name="vulnerabilities", remote_name="vulnerabilities")

        self._compute_args(**kwargs)

    def __str__(self):
        return '<%s:%s>' % (self.identity()["name"], self.identifier)

    def identifier(self):
        """ Identifier returns the value of the object's unique identifier.
        """
        return self.ID

    def setIdentifier(self, ID):
        """ SetIdentifier sets the value of the object's unique identifier.
        """
        self.ID = ID

    def identity(self):
        """ Identity returns the Identity of the object.
        """
        return layerIdentity

    # Properties
    @property
    def ID(self):
        """ Get ID value.

          Notes:
              ID is the identifier of the object.

              
        """
        return self._id

    @ID.setter
    def ID(self, value):
        """ Set ID value.

          Notes:
              ID is the identifier of the object.

              
        """
        self._id = value
    
    @property
    def name(self):
        """ Get name value.

          Notes:
              Name is the name of the entity

              
        """
        return self._name

    @name.setter
    def name(self, value):
        """ Set name value.

          Notes:
              Name is the name of the entity

              
        """
        self._name = value
    
    @property
    def namespaceName(self):
        """ Get namespaceName value.

          Notes:
              NamespaceName is the name of the namespace

              
        """
        return self._namespacename

    @namespaceName.setter
    def namespaceName(self, value):
        """ Set namespaceName value.

          Notes:
              NamespaceName is the name of the namespace

              
        """
        self._namespacename = value
    
    @property
    def parentName(self):
        """ Get parentName value.

          Notes:
              ParentName is the name of parent layer

              
        """
        return self._parentname

    @parentName.setter
    def parentName(self, value):
        """ Set parentName value.

          Notes:
              ParentName is the name of parent layer

              
        """
        self._parentname = value
    
    @property
    def severity(self):
        """ Get severity value.

          Notes:
              Severity defines the severity level of the vulnerability

              
        """
        return self._severity

    @severity.setter
    def severity(self, value):
        """ Set severity value.

          Notes:
              Severity defines the severity level of the vulnerability

              
        """
        self._severity = value
    
    @property
    def vulnerabilities(self):
        """ Get vulnerabilities value.

          Notes:
              Vulnerabilities is the list of all the vulnerabilities of a layer

              
        """
        return self._vulnerabilities

    @vulnerabilities.setter
    def vulnerabilities(self, value):
        """ Set vulnerabilities value.

          Notes:
              Vulnerabilities is the list of all the vulnerabilities of a layer

              
        """
        self._vulnerabilities = value
    
    def validate(self):
        """ Validate valides the current information stored into the structure.
        """
        errors = []

        err = validate_required_string("name", self.name)

        if err:
            errors.append(err)

        if len(errors) > 0:
            return errors

        return None

    # layerIdentity represents the Identity of the object
layerIdentity = {"name": "layer", "category": "layers", "constructor": Layer}

File: golang/layer.go
package gaia

import "fmt"
import "github.com/aporeto-inc/elemental"

import "github.com/aporeto-inc/gaia/golang/constants"
import "time"

// LayerIdentity represents the Identity of the object
var LayerIdentity = elemental.Identity{
	Name:     "layer",
	Category: "layers",
}

// LayersList represents a list of Layers
type LayersList []*Layer

// Layer represents the model of a layer
type Layer struct {
	// ID is the identifier of the object.
	ID string `json:"ID" cql:"id,omitempty" bson:"id"`

	// Vulnerability defines the overall vulnerability level of a layer
	Vulnerability constants.Vulnerability `json:"Vulnerability" cql:"vulnerability,omitempty" bson:"vulnerability"`

	// Annotation stores additional information about an entity
	Annotation map[string]string `json:"annotation" cql:"annotation,omitempty" bson:"annotation"`

	// AssociatedTags are the list of tags attached to an entity
	AssociatedTags []string `json:"associatedTags" cql:"associatedtags,omitempty" bson:"associatedtags"`

	// CreatedAt is the time at which an entity was created
	CreatedAt time.Time `json:"createdAt" cql:"createdat,omitempty" bson:"createdat"`

	// Deleted marks if the entity has been deleted.
	Deleted bool `json:"-" cql:"deleted,omitempty" bson:"deleted"`

	// Name is the name of the entity
	Name string `json:"name" cql:"name,primarykey,omitempty" bson:"_name"`

	// Namespace tag attached to an entity
	Namespace string `json:"namespace" cql:"namespace,primarykey,omitempty" bson:"_namespace"`

	// NamespaceName is the name of the namespace
	NamespaceName string `json:"namespaceName" cql:"namespacename,omitempty" bson:"namespacename"`

	// ParentID is the ID of the parent, if any,
	ParentID string `json:"parentID" cql:"parentid,omitempty" bson:"parentid"`

	// ParentName is the name of parent layer
	ParentName string `json:"parentName" cql:"parentname,omitempty" bson:"parentname"`

	// ParentType is the type of the parent, if any. It will be set to the parent's Identity.Name.
	ParentType string `json:"parentType" cql:"parenttype,omitempty" bson:"parenttype"`

	// Status of an entity
	Status constants.EntityStatus `json:"status" cql:"status,omitempty" bson:"status"`

	// UpdatedAt is the time at which an entity was updated.
	UpdatedAt time.Time `json:"updatedAt" cql:"updatedat,omitempty" bson:"updatedat"`
}

// NewLayer returns a new *Layer
func NewLayer() *Layer {

	return &Layer{
		Status: constants.Active,
	}
}

// Identity returns the Identity of the object.
func (o *Layer) Identity() elemental.Identity {

	return LayerIdentity
}

// Identifier returns the value of the object's unique identifier.
func (o *Layer) Identifier() string {

	return o.ID
}

func (o *Layer) String() string {

	return fmt.Sprintf("<%s:%s>", o.Identity().Name, o.Identifier())
}

// SetIdentifier sets the value of the object's unique identifier.
func (o *Layer) SetIdentifier(ID string) {

	o.ID = ID
}

// GetAssociatedTags returns the associatedTags of the receiver
func (o *Layer) GetAssociatedTags() []string {
	return o.AssociatedTags
}

// SetAssociatedTags set the given associatedTags of the receiver
func (o *Layer) SetAssociatedTags(associatedTags []string) {
	o.AssociatedTags = associatedTags
}

// SetCreatedAt set the given createdAt of the receiver
func (o *Layer) SetCreatedAt(createdAt time.Time) {
	o.CreatedAt = createdAt
}

// GetDeleted returns the deleted of the receiver
func (o *Layer) GetDeleted() bool {
	return o.Deleted
}

// SetDeleted set the given deleted of the receiver
func (o *Layer) SetDeleted(deleted bool) {
	o.Deleted = deleted
}

// GetNamespace returns the namespace of the receiver
func (o *Layer) GetNamespace() string {
	return o.Namespace
}

// SetNamespace set the given namespace of the receiver
func (o *Layer) SetNamespace(namespace string) {
	o.Namespace = namespace
}

// GetParentID returns the parentID of the receiver
func (o *Layer) GetParentID() string {
	return o.ParentID
}

// SetParentID set the given parentID of the receiver
func (o *Layer) SetParentID(parentID string) {
	o.ParentID = parentID
}

// GetParentType returns the parentType of the receiver
func (o *Layer) GetParentType() string {
	return o.ParentType
}

// SetParentType set the given parentType of the receiver
func (o *Layer) SetParentType(parentType string) {
	o.ParentType = parentType
}

// GetStatus returns the status of the receiver
func (o *Layer) GetStatus() constants.EntityStatus {
	return o.Status
}

// SetStatus set the given status of the receiver
func (o *Layer) SetStatus(status constants.EntityStatus) {
	o.Status = status
}

// SetUpdatedAt set the given updatedAt of the receiver
func (o *Layer) SetUpdatedAt(updatedAt time.Time) {
	o.UpdatedAt = updatedAt
}

// Validate valides the current information stored into the structure.
func (o *Layer) Validate() error {

	errors := elemental.Errors{}

	if len(errors) > 0 {
		return errors
	}

	return nil
}

// SpecificationForAttribute returns the AttributeSpecification for the given attribute name key.
func (o Layer) SpecificationForAttribute(name string) elemental.AttributeSpecification {

	return LayerAttributesMap[name]
}

// LayerAttributesMap represents the map of attribute for Layer.
var LayerAttributesMap = map[string]elemental.AttributeSpecification{
	"ID": elemental.AttributeSpecification{
		AllowedChoices: []string{},
		Autogenerated:  true,
		Exposed:        true,
		Filterable:     true,
		Format:         "free",
		Identifier:     true,
		Name:           "ID",
		Orderable:      true,
		ReadOnly:       true,
		Stored:         true,
		Type:           "string",
		Unique:         true,
	},
	"Vulnerability": elemental.AttributeSpecification{
		AllowedChoices: []string{},
		CreationOnly:   true,
		Exposed:        true,
		Name:           "Vulnerability",
		ReadOnly:       true,
		Required:       true,
		Stored:         true,
		SubType:        "vulnerability_level",
		Type:           "external",
	},
	"Annotation": elemental.AttributeSpecification{
		AllowedChoices: []string{},
		Exposed:        true,
		Name:           "annotation",
		Stored:         true,
		SubType:        "annotation",
		Type:           "external",
	},
	"AssociatedTags": elemental.AttributeSpecification{
		AllowedChoices: []string{},
		Exposed:        true,
		Getter:         true,
		Name:           "associatedTags",
		Setter:         true,
		Stored:         true,
		SubType:        "tags_list",
		Type:           "external",
	},
	"CreatedAt": elemental.AttributeSpecification{
		AllowedChoices: []string{},
		Autogenerated:  true,
		Exposed:        true,
		Filterable:     true,
		Name:           "createdAt",
		Orderable:      true,
		ReadOnly:       true,
		Setter:         true,
		Stored:         true,
		Type:           "time",
	},
	"Deleted": elemental.AttributeSpecification{
		AllowedChoices: []string{},
		Autogenerated:  true,
		Filterable:     true,
		Getter:         true,
		Name:           "deleted",
		Orderable:      true,
		Setter:         true,
		Stored:         true,
		Type:           "boolean",
	},
	"Name": elemental.AttributeSpecification{
		AllowedChoices: []string{},
		Exposed:        true,
		Filterable:     true,
		Format:         "free",
		Name:           "name",
		Orderable:      true,
		PrimaryKey:     true,
		Stored:         true,
		Type:           "string",
	},
	"Namespace": elemental.AttributeSpecification{
		AllowedChoices: []string{},
		Autogenerated:  true,
		CreationOnly:   true,
		Exposed:        true,
		Filterable:     true,
		Format:         "free",
		Getter:         true,
		Name:           "namespace",
		Orderable:      true,
		PrimaryKey:     true,
		ReadOnly:       true,
		Setter:         true,
		Stored:         true,
		Type:           "string",
		Unique:         true,
	},
	"NamespaceName": elemental.AttributeSpecification{
		AllowedChoices: []string{},
		Autogenerated:  true,
		CreationOnly:   true,
		Exposed:        true,
		Filterable:     true,
		Format:         "free",
		Name:           "namespaceName",
		Orderable:      true,
		ReadOnly:       true,
		Stored:         true,
		Type:           "string",
	},
	"ParentID": elemental.AttributeSpecification{
		AllowedChoices: []string{},
		Autogenerated:  true,
		Exposed:        true,
		Filterable:     true,
		ForeignKey:     true,
		Format:         "free",
		Getter:         true,
		Name:           "parentID",
		Orderable:      true,
		ReadOnly:       true,
		Setter:         true,
		Stored:         true,
		Type:           "string",
	},
	"ParentName": elemental.AttributeSpecification{
		AllowedChoices: []string{},
		Autogenerated:  true,
		CreationOnly:   true,
		Exposed:        true,
		Filterable:     true,
		Format:         "free",
		Name:           "parentName",
		Orderable:      true,
		ReadOnly:       true,
		Stored:         true,
		Type:           "string",
	},
	"ParentType": elemental.AttributeSpecification{
		AllowedChoices: []string{},
		Autogenerated:  true,
		Exposed:        true,
		Filterable:     true,
		Format:         "free",
		Getter:         true,
		Name:           "parentType",
		Orderable:      true,
		ReadOnly:       true,
		Setter:         true,
		Stored:         true,
		Type:           "string",
	},
	"Status": elemental.AttributeSpecification{
		AllowedChoices: []string{},
		Autogenerated:  true,
		Exposed:        true,
		Filterable:     true,
		Getter:         true,
		Name:           "status",
		Orderable:      true,
		Setter:         true,
		Stored:         true,
		SubType:        "status_enum",
		Type:           "external",
	},
	"UpdatedAt": elemental.AttributeSpecification{
		AllowedChoices: []string{},
		Autogenerated:  true,
		Exposed:        true,
		Filterable:     true,
		Name:           "updatedAt",
		Orderable:      true,
		ReadOnly:       true,
		Setter:         true,
		Stored:         true,
		Type:           "time",
	},
}
